File: PoketDex/app/src/main/java/pt/rfsfernandes/data/local/AppDatabase.java
package pt.rfsfernandes.data.local;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.sqlite.db.SupportSQLiteDatabase;
import pt.rfsfernandes.model.moves.Moves;
import pt.rfsfernandes.model.pokemon.Pokemon;
import pt.rfsfernandes.model.pokemon_species.PokemonSpecies;
import pt.rfsfernandes.model.service_responses.PokemonResult;

@Database(entities = {PokemonResult.class, Pokemon.class, PokemonSpecies.class, Moves.class},
    version = DBContract.DATABASE_VERSION,
    exportSchema = false)
public abstract class AppDatabase extends RoomDatabase {

  private static AppDatabase INSTANCE;

  public static AppDatabase getInstance(final Context context) {
    if (INSTANCE == null) {
      synchronized (AppDatabase.class) {
        if (INSTANCE == null) {
          INSTANCE = buildDatabase(context);
        }
      }
    }
    return INSTANCE;
  }

  private static AppDatabase buildDatabase(final Context context) {
    return Room.databaseBuilder(context.getApplicationContext(), AppDatabase.class, DBContract.DATABASE_NAME)
        .addCallback(new Callback() {
          @Override
          public void onCreate(@NonNull SupportSQLiteDatabase db) {
            super.onCreate(db);
          }
        })
        .build();
  }

  public abstract PokemonDAO getPokemonDAO();

}

File: PoketDex/app/src/main/java/pt/rfsfernandes/pocketdex/data/local/AppDatabase.kt
package pt.rfsfernandes.pocketdex.data.local

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import pt.rfsfernandes.pocketdex.model.moves.Moves
import pt.rfsfernandes.pocketdex.model.pokemon.Pokemon
import pt.rfsfernandes.pocketdex.model.pokemon_species.PokemonSpecies
import pt.rfsfernandes.pocketdex.model.service_responses.PokemonResult
import pt.rfsfernandes.pocketdex.model.type.Type

@Database(
    entities = [PokemonResult::class, Pokemon::class, PokemonSpecies::class, Moves::class, Type::class],
    version = DBContract.DATABASE_VERSION,
    exportSchema = false
)
abstract class AppDatabase : RoomDatabase() {
    abstract val pokemonDAO: PokemonDAO?

    companion object {
        private var INSTANCE: AppDatabase? = null
        fun getInstance(context: Context): AppDatabase? {
            if (INSTANCE == null) {
                synchronized(AppDatabase::class.java) {
                    if (INSTANCE == null) {
                        INSTANCE = buildDatabase(context)
                    }
                }
            }
            return INSTANCE
        }

        private fun buildDatabase(context: Context): AppDatabase {
            return Room.databaseBuilder(
                context.applicationContext,
                AppDatabase::class.java,
                DBContract.DATABASE_NAME
            )
                .addCallback(object : Callback() {
                })
                .build()
        }
    }
}